librz/il/rzil_vm.c

```diff
@@ -0,0 +1,460 @@
+	RzILBitVector *ret;
+	bool found = false;
+
+	rz_return_val_if_fail(lbl_name, NULL);
+	RzILEffectLabel *label = ht_pp_find(lbl_table, lbl_name, &found);
+	if (found) {
+		ret = label->addr;
+		return ret;
+	}
+	return NULL;
+}
+
+/**
+ * Find the label instance by name
+ * \param vm RzILVM, pointer to VM
+ * \param lbl_name string, the name of label
+ * \return lbl RzILEffectLabel, pointer to label instance
+ */
+RZ_API RzILEffectLabel *rz_il_vm_find_label_by_name(RzILVM *vm, RZ_NONNULL const char *lbl_name) {
+	rz_return_val_if_fail(lbl_name, NULL);
+	RzILEffectLabel *lbl = ht_pp_find(vm->vm_global_label_table, lbl_name, NULL);
+	return lbl;
+}
+
+/**
+ * Create a label in VM
+ * \param vm RzILVM, pointer to VM
+ * \param name string, name of label
+ * \param addr RzILBitVector, label address
+ * \return lbl RzILEffectLabel, pointer to label instance
+ */
+RZ_API RzILEffectLabel *rz_il_vm_create_label(RzILVM *vm, RZ_NONNULL const char *name, RzILBitVector *addr) {
+	rz_return_val_if_fail(name, NULL);
+	HtPP *lbl_table = vm->vm_global_label_table;
+
+	RzILEffectLabel *lbl = rz_il_effect_new_label(name, EFFECT_LABEL_ADDR);
+	lbl->addr = rz_il_bv_dup(addr);
+	ht_pp_insert(lbl_table, name, lbl);
+
+	return lbl;
+}
+
+/**
+ * Create a label without address, use rz_il_vm_update_label to update address for it
+ * \param vm RzILVM, pointer to VM
+ * \param name string, name of this label
+ * \return lbl RzILEffectLabel, pointer to label instance
+ */
+RZ_API RzILEffectLabel *rz_il_vm_create_label_lazy(RzILVM *vm, RZ_NONNULL const char *name) {
+	rz_return_val_if_fail(name, NULL);
+	HtPP *lbl_table = vm->vm_global_label_table;
+
+	RzILEffectLabel *lbl = rz_il_effect_new_label(name, EFFECT_LABEL_ADDR);
+	lbl->addr = NULL;
+	ht_pp_insert(lbl_table, name, lbl);
+
+	return lbl;
+}
+
+/**
+ * Update the address info of a label
+ * \param vm RzILVM, pointer to VM
+ * \param name string, name of this label
+ * \return lbl RzILEffectLabel, pointer to label instance
+ */
+RZ_API RzILEffectLabel *rz_il_vm_update_label(RzILVM *vm, RZ_NONNULL char *name, RzILBitVector *addr) {
+	rz_return_val_if_fail(name, NULL);
+	RzILEffectLabel *lbl = ht_pp_find(vm->vm_global_label_table, name, NULL);
+	if (lbl->addr) {
+		rz_il_bv_free(lbl->addr);
+	}
+	lbl->addr = rz_il_bv_dup(addr);
+
+	return lbl;
+}
+
+/**
+ * Store an opcode list to address
+ * \param vm RzILVM, pointer to VM
+ * \param addr RzILBitVector, address of this opcode list
+ * \param oplist RzPVector of RzILOp, core theory opcodes
+ */
+RZ_API void rz_il_vm_store_opcodes_to_addr(RzILVM *vm, RzILBitVector *addr, RzPVector *oplist) {
+	rz_return_if_fail(addr && oplist);
+	ht_pp_insert(vm->ct_opcodes, addr, oplist);
+}
+
+/**
+ * Make a core theory opcode vector
+ * \param num int, number of total opcodes
+ * \param ... op RzILOp, op will be pushed to vector one by one
+ * \return oplist RzPVector*, pointer to the opcode
+ */
+RZ_API RzPVector *rz_il_make_oplist(int num, ...) {
```

Comment:
This can be changed later, but I think it's less error prone to have NULL as terminator instead of having `num` as the number of following items.